src/ba.py

```diff
@@ -11,21 +11,37 @@ def border_array(x: str) -> list[int]:
     [0, 0, 0, 1, 2, 3, 4, 0, 0, 1]
     >>> border_array("")
     []
-    >>> strict_border_array("abaabaa")
+    >>> border_array("abaabaa")
     [0, 0, 1, 1, 2, 3, 4]
+    >>> border_array("abcabdabcabc")
+    [0, 0, 0, 1, 2, 0, 1, 2, 3, 4, 5, 3]
     """
     if x=="":
         return []
-    border_list = [0]
+    border_list = [0 for _ in x]
     index_match = 0
-    for c in x[1:]:
+    for i, c in enumerate(x[1:]):
         if c==x[index_match]:
             index_match += 1
-        elif c==x[0]:
-            index_match = 1
         else:
-            index_match = 0
-        border_list.append(index_match)
+            # See if we have another border inside the main one
+            inner_border = 0
+            start_inner_border = -1
+            j = i+1-index_match+1 # We start looking from the second character of the border
+            while j<i+2:
+                if x[j] == x[inner_border]:
+                    if inner_border == 0:
+                        start_inner_border = j
+                    inner_border += 1
+                else:
+                    if inner_border > 0:
+                        j = start_inner_border
+                        inner_border = 0
+                j += 1
+            
+            index_match = inner_border
```

Comment:
I have _no_ idea what is going on here. It is far more complicated than the algorithm I have shown you, and I am not convinced that this gives us a linear running time. It will take some convincing to get me there.

Reply:
I tried to do it by memory. I've looked for the solution and after following some examples of the algorithm by hand I think I understand it. I'm uploading the solution here